### ISSUE TITLE
[java] StringBuilder/Buffer false negatives with typeres

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#2881]
**Affects PMD Version:** 6.30.0-SNAPSHOT

**Rule:** [InsufficientStringBufferDeclaration](https://pmd.github.io/latest/pmd_rules_java_performance.html#insufficientstringbufferdeclaration),
[ConsecutiveLiteralAppends](https://pmd.github.io/latest/pmd_rules_java_performance.html#consecutiveliteralappends),
[AppendCharacterWithChar](https://pmd.github.io/latest/pmd_rules_java_performance.html#appendcharacterwithchar)

**Description:**

From https://github.com/pmd/pmd/pull/2871#issuecomment-717848659

**Code Sample demonstrating the issue:**

* https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A228
* https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/spring-framework/index.html#A1

```java
public class Foo {
    public void a() {
        StringBuilder sb = new StringBuilder();
        int start = sb.indexOf(" ");
        if (start == -1) {
            sb.append("Bar");
            sb.append(" "); // warning expected: Avoid appending characters as strings in StringBuffer.append. 
        }
    }
    public void b() {
        StringBuffer sb = new StringBuffer(); // warning expected: StringBuffer constructor is initialized with size 16, but has at least 17 characters appended. 
        sb.append("test "); // warning expected: StringBuffer (or StringBuilder).append is called 3 consecutive times with literals. Use a single append with a single combined String. 
        sb.append("test2 ").append("test3 ");
    }
}
```

**Expected outcome:**

*   Is PMD missing to report a violation, where there should be one? -> false-negative

**Running PMD through:** regression-tester

**Comment by oowekyala** [comment]
I wonder, if maybe #3061 helps with this

**Comment by pzygielo** [comment]
> I wonder, if maybe #3061 helps with this

Not sure if I was asked - but anyway - for my case PMD 6.31.0-SNAPSHOT (df80080aaa, built locally) - finds violations of `AppendCharacterWithChar` again where they are expected but are not recognized by 6.30. (And gives even more hits than 6.24.)

I will be happy to consume the fix in next release :grin:

**Comment by oowekyala** [comment]
Very nice, thanks for your input! Then, closing this issue could be just a matter of adding the relevant test cases to verify this is fixed

**Comment by pzygielo** [comment]
Thank you for the fix and brand new release.

> could be just a matter of adding the relevant test cases 

I tried to add some `test-code`s to `AppendCharacterWithChar.xml`, but with no success. I used the sample code from the first message here, and some extracts from my code I thought didn't trigger expected violation. UTs executed against 6.30 surprisingly passed. But I don't know the nature of the issue (yet :wink:) - I understand that my examples were not proper ones, and the added would pass on 6.31 and fail on 6.30.

**Comment by oowekyala** [comment]
Maybe it's because type resolution isn't properly configured in those tests. This issue was originally opened because the described regression happened when the auxclasspath was properly configured, but not otherwise. Originally it was spotted in our CI environment, I'm not sure if it has been reproduced elsewhere

You could try adding a copy of the test class into the java test sources tree. That way it's compiled with the rest of the test sources and the unit tests sees the class on the classpath

**Comment by oowekyala** [comment]
I added tests for this issue, which will be part of pmd 6.32.0. But I verified that the issue was fixed by #3061, so it was fixed in pmd 6.31.0
